contributors/design-proposals/storage/csi-migration.md

```diff
@@ -237,26 +237,48 @@ with the in-tree plugin, the VolumeAttachment object becomes orphaned.
-be modified to also check this location for a source before checking the PV
-itself.
+we will forward the in-tree volume source to CSI attach as-is and it will be
+copied to a new field in the VolumeAttachment object
+VolumeAttachment.Spec.Source.VolumeAttachmentSource.InlineVolumeSource. The
+VolumeAttachment name must be made with the CSI Translated version of the
+VolumeSource in order for it to be discoverable by Detach and WaitForAttach
+(described in more detail below).
+
+The CSI Attacher will have to be modified to also check this location for a
+source as well as checking for the `pvName`. Only one of the two may be
+specified.
+
+The new VolumeAttachmentSource API will look as such:
+```
+// VolumeAttachmentSource represents a volume that should be attached.
+// Right now only PersistenVolumes can be attached via external attacher,
+// in future we may allow also inline volumes in pods.
+// Exactly one member can be set.
+type VolumeAttachmentSource struct {
+	// Name of the persistent volume to attach.
+	// +optional
+	PersistentVolumeName *string `json:"persistentVolumeName,omitempty" protobuf:"bytes,1,opt,name=persistentVolumeName"`
+
+	// Allows CSI migration code to copy an inline volume
+	// source from a pod to the VolumeAttachment to support shimming of
+	// in-tree inline volumes to a CSI backend.
+	// This field is alpha-level and is only honored by servers that enable the CSIMigration feature.
+	// +optional
+	InlineVolumeSource *v1.VolumeSource `json:"inlineVolumeSource,omitempty protobuf:"bytes,2,opt,name=inlineVolumeSource"`
+}
+```
 
 We need to be careful with naming VolumeAttachments for in-line volumes. The
-name needs to be unique and ADC must be able to find the right
-VolumeAttachment when a pod is deleted (i.e. using only info in Node.Status).
-CSI driver in kubelet must be able to find the VolumeAttachment too to get
-AttachmentMetadata for NodeStage/NodePublish.
-
-In downgrade scenario where the migration is then turned off we will have to
-remove these floating VolumeAttachment objects, the same issue is outlined above
-in the Non-Dynamic Provisioned Volumes section.
-
-For more details on this see the PR that specs out CSI Inline Volumes in more detail:
-https://github.com/kubernetes/community/pull/2273. Basically we will just translate
-the in-tree inline volumes into the format specified/implemented in the 
-container-storage-interface-inline-volumes proposal.
+name needs to be unique and ADC must be able to find the right VolumeAttachment
+when a pod is deleted (i.e. using only info in Node.Status). CSI driver in
+kubelet must be able to find the VolumeAttachment too to call WaitForAttach and
+VolumesAreAttached.
+
+The attachment name is usually a hash of the volume name, CSI Driver name, and
+Node name. We are able to get all this information for Detach and WaitForAttach
+by translating the in-tree inline volume source to a CSI volume source before
+passing it to to the volume operations.
+
```

Comment:
do you want to add a note here that there's a current race condition for detach when a pod object is deleted and the AD controller restarts? And we can handle this in the future by reconciling the VolumeAttachment in the AD controller.

Reply:
actually thinking about this more, wont it get handled by the volume reconstruction code?

Reply:
reconstruction runs on the node, not for attacher

Reply:
i.e. for in-tree inline volumes reconstruction will go and clean up the mounts and detach. And once we have migration shim in volume reconstruction, it will do the proper volume attachment cleanup as well for migrated inline volumes

Reply:
added section for this